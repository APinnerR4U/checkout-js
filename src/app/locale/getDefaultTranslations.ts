import { Translations } from '@bigcommerce/checkout-sdk';

import { FALLBACK_TRANSLATIONS } from './translations';

export default async function getDefaultTranslations(locale: string): Promise<Translations> {
    switch (locale) {
        case 'de':
            return asTranslations((await import(
                /* webpackChunkName: "translations-de" */
                './translations/de.json'
            )).default);

        case 'fr':
            return asTranslations((await import(
                /* webpackChunkName: "translations-fr" */
                './translations/fr.json'
            )).default);

        case 'it':
            return asTranslations((await import(
                /* webpackChunkName: "translations-it" */
                './translations/it.json'
            )).default);

        case 'nl':
            return asTranslations((await import(
                /* webpackChunkName: "translations-nl" */
                './translations/nl.json'
            )).default);

        case 'pt-BR':
            return asTranslations((await import(
                /* webpackChunkName: "translations-pt-br" */
                './translations/pt-BR.json'
            )).default);

        case 'pt':
            return asTranslations((await import(
                /* webpackChunkName: "translations-pt" */
                './translations/pt.json'
            )).default);

        case 'sv':
            return asTranslations((await import(
                /* webpackChunkName: "translations-sv" */
                './translations/sv.json'
            )).default);

        case 'en':
            return FALLBACK_TRANSLATIONS;

        default:
            return {};
    }
}

function asTranslations(translations: unknown): Translations {
    return isTranslations(translations) ? translations : {};
}

function isTranslations(translations: unknown): translations is Translations {
    return typeof translations === 'object';
}
