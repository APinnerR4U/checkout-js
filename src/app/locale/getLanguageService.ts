import { createLanguageService, LanguageService } from '@bigcommerce/checkout-sdk';

import { memoize } from '../common/utility';
import defaultTranslations from '../language/en.json';

function getLanguageService(): LanguageService {
    return createLanguageService({
        ...(window as any).language,
        defaultTranslations,
    });
}

export default memoize(getLanguageService, { maxSize: 0 });
