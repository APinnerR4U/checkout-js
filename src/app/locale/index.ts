import { WithCurrencyProps } from './withCurrency';
import { WithLanguageProps } from './withLanguage';
import { LocaleContextType } from './LocaleContext';

export type LocaleContextType = LocaleContextType;
export type WithCurrencyProps = WithCurrencyProps;
export type WithLanguageProps = WithLanguageProps;

export { default as LocaleContext } from './LocaleContext';
export { default as createLocaleContext } from './createLocaleContext';
export { default as getLanguageService } from './getLanguageService';
export { default as withCurrency } from './withCurrency';
export { default as withLanguage } from './withLanguage';
export { default as LocaleProvider } from './LocaleProvider';
