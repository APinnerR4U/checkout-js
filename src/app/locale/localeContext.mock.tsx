import { createCurrencyService, createLanguageService } from '@bigcommerce/checkout-sdk';

import { getStoreConfig } from '../config/config.mock';
import defaultTranslations from '../language/en.json';

import { LocaleContextType } from './LocaleContext';

export function getLocaleContext(): Required<LocaleContextType> {
    return {
        currency: createCurrencyService(getStoreConfig()),
        language: createLanguageService({ ...(window as any).language, defaultTranslations }),
    };
}
