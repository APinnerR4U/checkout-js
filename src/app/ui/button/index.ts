import { ButtonProps } from './Button';

export type ButtonProps = ButtonProps;

export { default as Button, ButtonSize, ButtonVariant } from './Button';
