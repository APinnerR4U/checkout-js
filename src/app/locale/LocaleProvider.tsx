import { createCurrencyService, CheckoutService, LanguageService, StoreConfig } from '@bigcommerce/checkout-sdk';
import React, { Component, ReactNode } from 'react';

import getLanguageService from './getLanguageService';
import LocaleContext from './LocaleContext';

export interface LocaleProviderProps {
    checkoutService: CheckoutService;
}

export interface LocaleProviderState {
    config?: StoreConfig;
}

class LocaleProvider extends Component<LocaleProviderProps> {
    state: Readonly<LocaleProviderState>;

    private languageService: LanguageService;
    private unsubscribe?: () => void;

    constructor(props: Readonly<LocaleProviderProps>) {
        super(props);

        this.state = {};
        this.languageService = getLanguageService();
    }

    componentDidMount(): void {
        const { checkoutService } = this.props;

        this.unsubscribe = checkoutService.subscribe(
            ({ data }) => {
                this.setState({ config: data.getConfig() });
            },
            ({ data }) => data.getConfig()
        );
    }

    componentWillUnmount(): void {
        if (this.unsubscribe) {
            this.unsubscribe();
            this.unsubscribe = undefined;
        }
    }

    render(): ReactNode {
        const { children } = this.props;
        const { config } = this.state;
        const context = {
            currency: config ? createCurrencyService(config) : undefined,
            language: this.languageService,
        };

        return (
            <LocaleContext.Provider value={ context }>
                { children }
            </LocaleContext.Provider>
        );
    }
}

export default LocaleProvider;
