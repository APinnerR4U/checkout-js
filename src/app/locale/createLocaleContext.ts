import { createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';

import getLanguageService from './getLanguageService';
import { LocaleContextType } from './LocaleContext';

export default function createLocaleContext(config: StoreConfig): Required<LocaleContextType> {
    if (!config) {
        throw new Error('Missing configuration data');
    }

    return {
        currency: createCurrencyService(config),
        language: getLanguageService(),
    };
}
