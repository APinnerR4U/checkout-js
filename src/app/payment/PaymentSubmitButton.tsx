import React, { FunctionComponent } from 'react';

import { TranslatedString } from '../locale';
import { Button, ButtonSize, ButtonVariant } from '../ui/button';

import { PaymentMethodId, PaymentMethodType } from './paymentMethod';

interface PaymentSubmitButtonTextProps {
    methodId?: string;
    methodType?: string;
}

const PaymentSubmitButtonText: FunctionComponent<PaymentSubmitButtonTextProps> = ({ methodId, methodType }) => {
    if (methodId === PaymentMethodId.Amazon) {
        return <TranslatedString id="payment.amazon_continue_action" />;
    }

    if (methodType === PaymentMethodType.VisaCheckout) {
        return <TranslatedString id="payment.visa_checkout_continue_action" />;
    }

    if (methodType === PaymentMethodType.Chasepay) {
        return <TranslatedString id="payment.chasepay_continue_action" />;
    }

    if (methodType === PaymentMethodType.Paypal) {
        return <TranslatedString id="payment.paypal_continue_action" />;
    }

    if (methodType === PaymentMethodType.PaypalCredit) {
        return <TranslatedString id="payment.paypal_credit_continue_action" />;
    }

    return <TranslatedString id="payment.place_order_action" />;
};

interface PaymentSubmitButtonProps {
    isDisabled?: boolean;
    isLoading?: boolean;
    methodId?: string;
    methodType?: string;
}

const PaymentSubmitButton: FunctionComponent<PaymentSubmitButtonProps> = ({
    isLoading,
    isDisabled,
    methodId,
    methodType,
}) => (
    <Button
        disabled={ isDisabled }
        id="checkout-payment-continue"
        isFullWidth
        isLoading={ isLoading }
        size={ ButtonSize.Large }
        type="submit"
        variant={ ButtonVariant.Action }
    >
        <PaymentSubmitButtonText
            methodId={ methodId }
            methodType={ methodType }
        />
    </Button>
);

export default PaymentSubmitButton;
