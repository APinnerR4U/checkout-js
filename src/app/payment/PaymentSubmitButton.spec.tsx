import { LanguageService } from '@bigcommerce/checkout-sdk';
import { mount, shallow } from 'enzyme';
import React from 'react';

import { getStoreConfig } from '../config/config.mock';
import { createLocaleContext, LocaleContext, LocaleContextType } from '../locale';
import { Button } from '../ui/button';

import PaymentSubmitButton from './PaymentSubmitButton';

describe('PaymentSubmitButton', () => {
    let localeContext: LocaleContextType;
    let languageService: LanguageService;

    beforeEach(() => {
        localeContext = createLocaleContext(getStoreConfig());
        languageService = localeContext.language;
    });

    it('matches snapshot with rendered output', () => {
        const component = shallow(<PaymentSubmitButton />);

        expect(component)
            .toMatchSnapshot();
    });

    it('forwards props to button', () => {
        const component = shallow(
            <PaymentSubmitButton
                isDisabled
                isLoading
            />
        );

        expect(component.find(Button).props())
            .toEqual(expect.objectContaining({
                disabled: true,
                isLoading: true,
            }));
    });

    it('renders button with default label', () => {
        const component = mount(
            <LocaleContext.Provider value={ localeContext }>
                <PaymentSubmitButton />
            </LocaleContext.Provider>
        );

        expect(component.text())
            .toEqual(languageService.translate('payment.place_order_action'));
    });

    it('renders button with special label for Amazon', () => {
        const component = mount(
            <LocaleContext.Provider value={ localeContext }>
                <PaymentSubmitButton methodId="amazon" />
            </LocaleContext.Provider>
        );

        expect(component.text())
            .toEqual(languageService.translate('payment.amazon_continue_action'));
    });

    it('renders button with special label for Visa Checkout provided by Braintree', () => {
        const component = mount(
            <LocaleContext.Provider value={ localeContext }>
                <PaymentSubmitButton methodType="visa-checkout" />
            </LocaleContext.Provider>
        );

        expect(component.text())
            .toEqual(languageService.translate('payment.visa_checkout_continue_action'));
    });

    it('renders button with special label for ChasePay', () => {
        const component = mount(
            <LocaleContext.Provider value={ localeContext }>
                <PaymentSubmitButton methodType="chasepay" />
            </LocaleContext.Provider>
        );

        expect(component.text())
            .toEqual(languageService.translate('payment.chasepay_continue_action'));
    });

    it('renders button with special label for PayPal', () => {
        const component = mount(
            <LocaleContext.Provider value={ localeContext }>
                <PaymentSubmitButton methodType="paypal" />
            </LocaleContext.Provider>
        );

        expect(component.text())
            .toEqual(languageService.translate('payment.paypal_continue_action'));
    });

    it('renders button with special label for PayPal Credit', () => {
        const component = mount(
            <LocaleContext.Provider value={ localeContext }>
                <PaymentSubmitButton methodType="paypal-credit" />
            </LocaleContext.Provider>
        );

        expect(component.text())
            .toEqual(languageService.translate('payment.paypal_credit_continue_action'));
    });
});
