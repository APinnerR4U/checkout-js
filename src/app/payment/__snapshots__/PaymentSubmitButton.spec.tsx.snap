// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentSubmitButton matches snapshot with rendered output 1`] = `
<Button
  id="checkout-payment-continue"
  isFullWidth={true}
  size="large"
  type="submit"
  variant="action"
>
  <PaymentSubmitButtonText />
</Button>
`;
